### PR TITLE
Windows manager

### PR DESCRIPTION
**Pull request by andrzejgorski**

### CODE REVIEW

File: whylog_vim/gui/windows.py [hunk 1]
```diff
@@ -46,3 +51,55 @@ def get_window_id(self):
         if window_id:
             return window_id
         raise CannotFindWindowId(self.name)
+
+
+class WhylogWindowManager(object):
+
+    whylog_windows = [WindowTypes.QUERY, WindowTypes.TEACHER, WindowTypes.INPUT, WindowTypes.CASE,]
```
Comment (kgromadzki) [inline whylog_vim/gui/windows.py]: Use tuple instead of list. You have a guarantee of immutability.

File: whylog_vim/gui/windows.py [hunk 2]
```diff
@@ -46,3 +51,55 @@ def get_window_id(self):
...
+        return content
+
+    def go_to_window(self, window_type):
+        try:
+            VimUIWrapper.go_to_window(self.windows[window_type].get_window_id())
+        except KeyError:
+            raise CannotSwitchToWindow(window_type)
+
+    def get_window_id(self, window_type):
+        try:
```
Comment (kgromadzki) [inline whylog_vim/gui/windows.py]: Refactor to:

``` python
window = self.windows.get(window_type)
if window is not None:
    return window.get_window_id()
```

File: whylog_vim/gui/windows.py [hunk 3]
```diff
@@ -46,3 +51,55 @@ def get_window_id(self):
...
+        except CannotSwitchToWindow:
+            raise CannotCloseWindow(window_type)
+        VimUIWrapper.close_current_window()
+        del self.windows[window_type]
+
+    def are_windows_closed(self):
+        return not self.windows
+
+    def get_windows_ids(self):
+        return [window.get_window_id() for window in self.windows.values()]
```
Comment (kgromadzki) [inline whylog_vim/gui/windows.py]: Use six.itervalues

File: whylog_vim/gui/windows.py [hunk 4]
```diff
@@ -46,3 +51,55 @@ def get_window_id(self):
...
+
+    def set_content(self, window_type, content):
+        try:
+            self.windows[window_type].set_content(content)
+        except KeyError:
+            CannotSetWindowContent(window_type)
```
Comment (kgromadzki) [inline whylog_vim/gui/windows.py]: Do you want to raise an exception? If so, raise it.

File: whylog_vim/gui/windows.py [hunk 5]
```diff
@@ -46,3 +53,52 @@ def get_window_id(self):
         if window_id:
             return window_id
         raise CannotFindWindowId(self.name)
+
+
+class WhylogWindowManager(object):
+
+    whylog_windows = (WindowTypes.QUERY, WindowTypes.TEACHER, WindowTypes.INPUT, WindowTypes.CASE)
```
Comment (ppolewicz) [inline whylog_vim/gui/windows.py]: `WHYLOG_WINDOWS = ...`

Comment (ppolewicz) [inline whylog_vim/gui/windows.py]: what is this for? seems not to be used

File: whylog_vim/gui/windows.py [hunk 6]
```diff
@@ -46,3 +53,52 @@ def get_window_id(self):
...
+
+    def __init__(self):
+        self.windows = dict()
+
+    def create_window(self, window_type, content, modifiable=False, size=None):
+        self.windows[window_type] = Window(window_type, content, modifiable, size)
+
+    def get_window_content(self, window_type):
+        try:
```
Comment (ppolewicz) [inline whylog_vim/gui/windows.py]: to DRY, make a decorator like this:

``` python
try:
    function(window_type, *args, **kwargs)
except KeyError:
    raise CannotSwitchToWindow(window_type)
```

**Pull request merged**